### DISCUSSION TITLE
Which library(ct-lib/db-lib) does the app should using for MS SQLServer?

### DISCUSSION THREAD

**Discussion by taozuhong** [FreeTDS/freetds#510] [General]
> The code is split into several pieces.
> 1. `tds` directory is the wire level stuff, it should be independent of the library using it, this will allow db-lib, ct-lib, and ODBC to sit on top.
> 2. `db-lib` directory. This is the actual db-lib code which runs on top of tds.
> 3. `ct-lib` directory. This is the ct-lib code which runs on top of tds.

I want to support Microsoft SQL Server, which library(ct-lib/db-lib) does the app should using?

**Comment by taozuhong** [comment]
The prefer library for Microsoft SQL Server is `db-lib`, please see the guide:
https://www.freetds.org/userguide/projects.html

[TinyTDS](https://github.com/rails-sqlserver/tiny_tds) - Simple and fast FreeTDS bindings for Ruby using **DB-Library**.
https://github.com/rails-sqlserver/tiny_tds

**Comment by freddy77** [comment]
Why not ODBC ?

**Comment by taozuhong** [comment]
App could connect to server via db-lib, why ODBC? 

It will introduce more bugs and more layers and more complications.

**Reply by freddy77** [reply]
The problem with db-lib is that some specific features are missing (Microsoft stopped developing it many years ago). Sure, support on FreeTDS is better, you can use UTF-8 so you won't miss wide characters support but things like MARS, TVP, query notifications and others will be always missing. Potentially you can use `--with-odbc-nodm` option and avoid the Driver Manager layer although this remove some check DM does that you'll have to check on yourself. Obviously, if you don't need additional features and db-lib is easier for you use db-lib.
